### PR TITLE
[Tizen] Add url decoding for 'uri' attribute of signatures

### PR DESCRIPTION
**Pull request by sung-su**
The value of URI attribute of signature files can be url encoded string.
But, SignatureValidator can't recongize url encoded 'uri' of signatures.
Add url decoding routine to SignatureParser.

BUG=XWALK-3479

### CODE REVIEW

File: application/common/tizen/signature_parser.cc [hunk 1]
```diff
@@ -111,6 +111,34 @@ xmlNsPtr GetNamespace(const xmlNodePtr node, const char* expected_href) {
 
 namespace xwalk {
 namespace application {
```
Comment (halton) [inline application/common/tizen/signature_parser.cc]: nit, empty line after

File: application/common/tizen/signature_parser.cc [hunk 2]
```diff
@@ -111,6 +111,34 @@ xmlNsPtr GetNamespace(const xmlNodePtr node, const char* expected_href) {
...
+int hexToInt(char a) {
```
Comment (halton) [inline application/common/tizen/signature_parser.cc]: Please follow chromium naming rule for function, `hexToInt` -> `HexToInt`

File: application/common/tizen/signature_parser.cc [hunk 3]
```diff
@@ -111,6 +111,34 @@ xmlNsPtr GetNamespace(const xmlNodePtr node, const char* expected_href) {
...
+  if (a >= '0' && a <= '9') return a - '0';
+  if (a >= 'A' && a <= 'F') return a - 'A' + 10;
+  if (a >= 'a' && a <= 'f') return a - 'a' + 10;
+  return -1;
```
Comment (halton) [inline application/common/tizen/signature_parser.cc]: I'd suggest to copy the implemation of [src/third_party/WebKit/Source/core/inspector/JSONParser.cpp] (https://code.google.com/p/chromium/codesearch#chromium/src/third_party/WebKit/Source/core/inspector/JSONParser.cpp&q=hextoInt&sq=package:chromium&l=258&ct=rc&cd=6&dr=CSs). 

Note, please change to chromium coding style, not the WebKit/Blink style.

File: application/common/tizen/signature_parser.cc [hunk 4]
```diff
@@ -111,6 +111,34 @@ xmlNsPtr GetNamespace(const xmlNodePtr node, const char* expected_href) {
...
+}
+
+std::string decodeProcent(const std::string& path) {
```
Comment (halton) [inline application/common/tizen/signature_parser.cc]: A example to explain how the decoding look like should be helpful

File: application/common/tizen/signature_parser.cc [hunk 5]
```diff
@@ -151,6 +179,7 @@ bool ParseSignedInfoElement(
   }
 
   std::string uri;
+  std::string decoded_uri;
```
Comment (halton) [inline application/common/tizen/signature_parser.cc]: declare this var as close as first time used.

File: application/common/tizen/signature_parser.cc (hunk 2 above)
Comment (halton) [inline application/common/tizen/signature_parser.cc]: Look the code again, please  put `HexToInt()` and `DecodeProcent()` under anonymous namespace.

**Comment by rakuco** [comment]
this kind of parsing is exactly the kind of code I'd expect to find in Chromium. did you look for it before rolling your own version?

**Comment by halton** [comment]
> this kind of parsing is exactly the kind of code I'd expect to find in Chromium. did you look for it before rolling your own version?

I did rough searching chromium code, all the HexToInt is private. Anyway @sung-su can do more searching.

**Comment by sung-su** [comment]
I've fixed except "DecodeProcent" function.
I found the decode function in the Chromium base source code which has similar functionality. However, it is not suitable because complex.

File: application/common/tizen/signature_parser.cc [hunk 6]
```diff
@@ -160,6 +185,12 @@ bool ParseSignedInfoElement(
       LOG(ERROR) << "Missing URI attribute.";
       return false;
     }
+    std::string decoded_uri = DecodeProcent(uri);
+    if (!decoded_uri.empty()) {
+      uri = decoded_uri;
+    } else {
+      LOG(WARNING) << "Failing to decode URI.";
```
Comment (rakuco) [inline application/common/tizen/signature_parser.cc]: If decoding fails, shouldn't you abort whatever you're doing and fail?

**Comment by rakuco** [comment]
> However, it is not suitable because complex.

That doesn't sound like a convincing argument. Writing your own URI decoding function should be your very last resort. Doesn't `url::DecodeURLEscapeSequences` solve your problem?
Also, have you checked if other the parsing of other URI fields also have this problem?

**Comment by sung-su** [comment]
I've fixed.
@rakuco 
If decoding fails, abort at the validation check step.

**Comment by pozdnyakov** [comment]
lgtm

**Pull request merged**